gems/quilt_rails/README.md

```diff
@@ -215,18 +217,205 @@ class ReactController < ApplicationController
+
+The `Quilt::Performance::Client` class is yielded into the block parameter for `process_report`, and is the primary API for customizing what metrics are sent for a given POST.
+
+##### Client#on_distribution
+
+The `on_distribution` method takes a block which is run for each distribution (including custom ones) sent during `process_report`. It can be used to easily add logging or other side-effects to your measurements.
+
+```ruby
+client.on_distribution do |metric_name, value, tags|
+  Rails.logger.debug "#{metric_name}: #{value}, tags: #{tags}"
+end
+```
+
+##### Client#on_distribution
+
+The `on_distribution` method takes a block which is run for each distribution (including custom ones) sent during `process_report`. It can be used to easily add logging or other side-effects to your measurements.
+
+```ruby
+client.on_distribution do |metric_name, value, tags|
+  Rails.logger.debug "#{metric_name}: #{value}, tags: #{tags}"
+end
+```
+
+##### Client#on_report
+
+The `on_report` method takes a block which is run once per report POSTed to the controller. It can be used to perform side-effects based on the entire `Report`, including sending custom `distribution`s.
+
+```ruby
+client.on_report do |report, tags|
+  client.distribution('some_metric', some_computation(report), tags)
+end
+```
+
+##### Client#on_navigation
+
+The `on_report` method takes a block which is run once per navigation in a given `Report`. It can be used to perform side-effects based on a given `Navigation`, including sending custom `distribution`s.
+
+```ruby
+client.on_navigation do |navigation, tags|
+  weight = navigation.events_with_size.reduce(0) do |total, event|
+    total + event.size
+  end
+
+  client.distribution('navigation_total_resource_weight', weight, tags)
+end
+```
+
+##### Client#customize_tags
+
+The `customize_tags` method allows custom tags to be computed to include in every distribution for a given POST to the controller.
+
+```ruby
+client.customize_tags do |metrics_body, tags|
+  # pathname will be inluded on *every* distribution for this report
+  tags[:pathname] = metrics_body.pathname
+end
+```
+
+##### Client#customize_navigation_tags
+
+The `customize_navigation_tags` method allows custom tags to be computed for each individual navigation described by the POST body, and included on the `distribution` calls for that navigation.
+
+```ruby
+client.customize_navigation_tags do |navigation, tags|
+  # target will be included only for the distributions about this navigation
+  tags[:navigation_target] = navigation.target
+end
+```
+
+#### Report
+
+[`Quilt::Performance::Report`](/lib/quilt_rails/performance/report.rb#L3) describes timing, size information, and metadata about a series of events and navigations on the front-end of a web application, as recorded by `@shopify/performance` and parsed from an `ActionController:Parameters` instance.
+
+##### Report#pathname
+
+The initial path (eg. `/foo`) of the request which generated the report.
```

Comment:
Does this even make sense if it contains multiple navigations?

Reply:
That's a good question, probably not. Should we delete our usage of `pathname` from web and koa-metrics?

Reply:
Thinking about it more, it makes sense specifically as the pathname that the top level events (such as TTFB) refer to. `Navigation`s are just client-side `History` style navigations. There's no way for batching to even happen across multiple server side requests.